File: app/src/main/java/org/nitri/opentopo/WayPointDetailDialogFragment.java
package org.nitri.opentopo;

import android.annotation.SuppressLint;
import android.app.Dialog;
import android.os.Bundle;
import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;
import androidx.appcompat.app.AlertDialog;
import android.text.method.LinkMovementMethod;
import android.view.LayoutInflater;
import android.view.View;
import android.view.Window;
import android.widget.TextView;

import org.nitri.opentopo.model.WayPointItem;


public class WayPointDetailDialogFragment extends DialogFragment {

    private Callback mCallback;

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {

        if (requireActivity().getSupportFragmentManager() != null)
            mCallback = (Callback) requireActivity().getSupportFragmentManager().findFragmentByTag(MainActivity.GPX_DETAIL_FRAGMENT_TAG);

        AlertDialog.Builder builder = new AlertDialog.Builder(requireActivity());
        LayoutInflater inflater = requireActivity().getLayoutInflater();

        // Pass null as the parent view because it's going in the dialog layout
        @SuppressLint("InflateParams")
        View rootView = inflater.inflate(R.layout.fragment_way_point_detail, null);

        TextView tvName = rootView.findViewById(R.id.tvTitle);
        TextView tvDescription = rootView.findViewById(R.id.tvDescription);

        tvDescription.setMovementMethod(LinkMovementMethod.getInstance());

        builder.setView(rootView);

        Dialog dialog = builder.create();
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE);

        if (mCallback != null) {
            WayPointItem item = mCallback.getSelectedWayPointItem();
            if (item != null && item.getWayPoint() != null) {
                tvName.setText(item.getWayPoint().getName());
                tvDescription.setText(Util.fromHtml(item.getWayPoint().getDesc().replace("href=\"//", "href=\"http://")));
            }
        }

        return dialog;
    }


    interface Callback {
        WayPointItem getSelectedWayPointItem();
    }
}

File: app/src/main/java/org/nitri/opentopo/WayPointDetailDialogFragment.kt
package org.nitri.opentopo

import android.app.Dialog
import android.os.Bundle
import android.view.Window
import androidx.appcompat.app.AlertDialog
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.ComposeView
import androidx.compose.ui.platform.ViewCompositionStrategy
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.unit.dp
import androidx.fragment.app.DialogFragment
import androidx.lifecycle.setViewTreeLifecycleOwner
import androidx.lifecycle.setViewTreeViewModelStoreOwner
import androidx.savedstate.setViewTreeSavedStateRegistryOwner
import org.nitri.opentopo.model.WayPointItem
import org.nitri.opentopo.ui.theme.OpenTopoTheme
import org.nitri.opentopo.view.HtmlText

class WayPointDetailDialogFragment : DialogFragment() {
    private var mCallback: Callback? = null

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val fragmentActivity = requireActivity()
        mCallback =
            fragmentActivity.supportFragmentManager.findFragmentByTag(
                BaseMainActivity.GPX_DETAIL_FRAGMENT_TAG
            ) as Callback?

        val wayPoint = mCallback?.getSelectedWayPointItem()?.wayPoint
        val name = wayPoint?.name.orEmpty()
        val description = wayPoint?.desc
            ?.takeIf { it.isNotBlank() }
            ?.replace("href=\"//", "href=\"http://")

        val composeView = ComposeView(fragmentActivity).apply {
            setViewCompositionStrategy(ViewCompositionStrategy.DisposeOnViewTreeLifecycleDestroyed)
            setContent {
                OpenTopoTheme(dynamicColor = false) {
                    WayPointDetailContent(
                        name = name,
                        description = description
                    )
                }
            }
        }

        composeView.setViewTreeLifecycleOwner(this)
        composeView.setViewTreeViewModelStoreOwner(this)
        composeView.setViewTreeSavedStateRegistryOwner(this)

        val dialog = AlertDialog.Builder(fragmentActivity)
            .setView(composeView)
            .create()
        dialog.requestWindowFeature(Window.FEATURE_NO_TITLE)

        dialog.setOnShowListener {
            dialog.window?.decorView?.let { decorView ->
                decorView.setViewTreeLifecycleOwner(this)
                decorView.setViewTreeViewModelStoreOwner(this)
                decorView.setViewTreeSavedStateRegistryOwner(this)
            }
        }

        return dialog
    }

    internal interface Callback {
        fun getSelectedWayPointItem(): WayPointItem?
    }
}

@Composable
private fun WayPointDetailContent(
    name: String,
    description: String?
) {
    Column(
        modifier = Modifier
            .fillMaxWidth()
            .verticalScroll(rememberScrollState())
            .padding(16.dp),
        horizontalAlignment = Alignment.CenterHorizontally
    ) {
        Text(
            text = name,
            modifier = Modifier.fillMaxWidth(),
            style = MaterialTheme.typography.titleLarge,
            textAlign = TextAlign.Center
        )
        if (description != null) {
            HtmlText(
                html = description,
                modifier = Modifier
                    .fillMaxWidth()
                    .padding(top = 8.dp)
            )
        }
    }
}
